pkg/resourcemanager/webhook/podzoneaffinity/handler.go

```diff
@@ -0,0 +1,224 @@
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+var podGVK = metav1.GroupVersionKind{Group: "", Kind: "Pod", Version: "v1"}
+
+type handler struct {
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+// NewHandler returns a new handler.
+func NewHandler() admission.Handler {
+	return &handler{}
+}
+
+func (h *handler) InjectClient(cl client.Client) error {
+	h.client = cl
+	return nil
+}
+
+func (h *handler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if req.Operation != admissionv1.Create {
+		return admission.Allowed("only 'create' operation is handled")
+	}
+
+	if req.Kind != podGVK {
+		return admission.Allowed("resource is not corev1.Pod")
+	}
+
+	if req.SubResource != "" {
+		return admission.Allowed("subresources on pods are not supported")
+	}
+
+	pod := &corev1.Pod{}
+	if err := h.decoder.Decode(req, pod); err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	}
+
+	// Check conflicting and add required pod affinity terms.
+	handlePodAffinity(pod)
+
+	// If the concrete zone is already determined by Gardener, let the pod be scheduled only to nodes in that zone.
+	if err := handleNodeAffinity(ctx, h.client, pod); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	marshaledPod, err := json.Marshal(pod)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+}
+
+func handlePodAffinity(pod *corev1.Pod) {
+	affinityTerm := corev1.PodAffinityTerm{
+		// Use empty label selector to match any pods in the shoot control plane namespace.
+		LabelSelector: &metav1.LabelSelector{},
+		TopologyKey:   corev1.LabelTopologyZone,
+	}
+
+	// First remove potentially conflicting pod anti affinity terms that would forbid scheduling into a single zone.
+	removeConflictingPodAntiAffinityTerms(pod, affinityTerm)
+
+	// Handle required affinity to let the pod be scheduled to a specific zone.
+	handleZonePodAffinityTerm(pod, affinityTerm)
+}
+
+func removeConflictingPodAntiAffinityTerms(pod *corev1.Pod, affinityTerm corev1.PodAffinityTerm) {
+	if pod.Spec.Affinity == nil || pod.Spec.Affinity.PodAntiAffinity == nil {
+		return
+	}
+
+	// Remove pod anti-affinity rules with zone topology key.
+	remainingAntiAffinityTerms := make([]corev1.PodAffinityTerm, 0, len(pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution))
+	for _, term := range pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+		if equality.Semantic.DeepEqual(term, affinityTerm) {
+			continue
```

Comment:
I have never considered that option. It's surely good for debugging and troubleshooting. A log entry could also do the job. Any opinions @rfranzke?

Reply:
A log entry would be good, yes. I would not add an annotation. However, the documentation already states that this might happen, so it shouldn't be surprising to operators (all this is only happening in seeds).